core/preconditioner/isai_kernels.hpp

```diff
@@ -51,29 +51,41 @@ namespace kernels {
                               IndexType *excess_rhs_ptrs,                     \
                               IndexType *excess_nz_ptrs, bool lower)
 
+
+#define GKO_DECLARE_ISAI_GENERATE_GENERAL_INVERSE_KERNEL(ValueType, IndexType) \
+    void generate_general_inverse(                                             \
+        std::shared_ptr<const DefaultExecutor> exec,                           \
+        const matrix::Csr<ValueType, IndexType> *input,                        \
+        matrix::Csr<ValueType, IndexType> *inverse,                            \
+        IndexType *excess_rhs_ptrs, IndexType *excess_nz_ptrs)
+
 #define GKO_DECLARE_ISAI_GENERATE_EXCESS_SYSTEM_KERNEL(ValueType, IndexType) \
     void generate_excess_system(                                             \
         std::shared_ptr<const DefaultExecutor> exec,                         \
         const matrix::Csr<ValueType, IndexType> *input,                      \
         const matrix::Csr<ValueType, IndexType> *inverse,                    \
         const IndexType *excess_rhs_ptrs, const IndexType *excess_nz_ptrs,   \
         matrix::Csr<ValueType, IndexType> *excess_system,                    \
-        matrix::Dense<ValueType> *excess_rhs)
+        matrix::Dense<ValueType> *excess_rhs, const size_type excess_start,  \
+        const size_type excess_end)
 
 #define GKO_DECLARE_ISAI_SCATTER_EXCESS_SOLUTION_KERNEL(ValueType, IndexType) \
     void scatter_excess_solution(                                             \
         std::shared_ptr<const DefaultExecutor> exec,                          \
         const IndexType *excess_rhs_ptrs,                                     \
         const matrix::Dense<ValueType> *excess_solution,                      \
-        matrix::Csr<ValueType, IndexType> *inverse)
-
-#define GKO_DECLARE_ALL_AS_TEMPLATES                                      \
-    constexpr auto row_size_limit = 32;                                   \
-    template <typename ValueType, typename IndexType>                     \
-    GKO_DECLARE_ISAI_GENERATE_TRI_INVERSE_KERNEL(ValueType, IndexType);   \
-    template <typename ValueType, typename IndexType>                     \
-    GKO_DECLARE_ISAI_GENERATE_EXCESS_SYSTEM_KERNEL(ValueType, IndexType); \
-    template <typename ValueType, typename IndexType>                     \
+        matrix::Csr<ValueType, IndexType> *inverse, const size_type e_start,  \
+        const size_type e_end)
```

Comment:
nit: `const` is an implementation detail and not part of the interface
```suggestion
        matrix::Csr<ValueType, IndexType> *inverse, size_type e_start,        \
        size_type e_end)
```